File: app/src/main/java/info/nightscout/androidaps/plugins/PumpDanaRS/comm/DanaRS_Packet_Option_Set_User_Option.java
package info.nightscout.androidaps.plugins.PumpDanaRS.comm;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.nightscout.androidaps.Config;
import com.cozmo.danar.util.BleCommandUtil;
import info.nightscout.androidaps.plugins.PumpDanaR.DanaRPump;

public class DanaRS_Packet_Option_Set_User_Option extends DanaRS_Packet {
    private static Logger log = LoggerFactory.getLogger(DanaRS_Packet_Option_Set_User_Option.class);

    private int error;

    public DanaRS_Packet_Option_Set_User_Option() {
        super();
        opCode = BleCommandUtil.DANAR_PACKET__OPCODE_OPTION__SET_USER_OPTION;
        if (Config.logDanaMessageDetail) {
            log.debug("Setting user settings");
        }
    }

    @Override
    public byte[] getRequestParams() {
        DanaRPump pump = DanaRPump.getInstance();

        byte[] request = new byte[13];
        request[0] = (byte) (pump.timeDisplayType & 0xff);
        request[1] = (byte) (pump.buttonScrollOnOff & 0xff);
        request[2] = (byte) (pump.beepAndAlarm & 0xff);
        request[3] = (byte) (pump.lcdOnTimeSec & 0xff);
        request[4] = (byte) (pump.backlightOnTimeSec & 0xff);
        request[5] = (byte) (pump.selectedLanguage & 0xff);
        request[6] = (byte) (pump.units & 0xff);
        request[7] = (byte) (pump.shutdownHour & 0xff);
        request[8] = (byte) (pump.lowReservoirRate & 0xff);
        request[9] = (byte) (pump.cannulaVolume & 0xff);
        request[10] = (byte) ((pump.cannulaVolume >>> 8) & 0xff);
        request[11] = (byte) (pump.refillAmount & 0xff);
        request[12] = (byte) ((pump.refillAmount >>> 8) & 0xff);
        return request;
    }

    @Override
    public void handleMessage(byte[] data) {
        int dataIndex = DATA_START;
        int dataSize = 1;
        error = byteArrayToInt(getBytes(data, dataIndex, dataSize));
        if (Config.logDanaMessageDetail) {
            if (error == 0)
                log.debug("Result OK");
            else
                log.error("Result Error: " + error);
        }
    }

    @Override
    public String getFriendlyName() {
        return "OPTION__SET_USER_OPTION";
    }
}

File: danars/src/main/java/info/nightscout/androidaps/danars/comm/DanaRS_Packet_Option_Set_User_Option.kt
package info.nightscout.androidaps.danars.comm

import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.logging.LTag
import info.nightscout.androidaps.dana.DanaPump
import info.nightscout.androidaps.danars.encryption.BleEncryption
import javax.inject.Inject

class DanaRS_Packet_Option_Set_User_Option(
    injector: HasAndroidInjector
) : DanaRS_Packet(injector) {

    @Inject lateinit var danaPump: DanaPump

    init {
        opCode = BleEncryption.DANAR_PACKET__OPCODE_OPTION__SET_USER_OPTION
        aapsLogger.debug(LTag.PUMPCOMM, "Setting user settings")
    }

    override fun getRequestParams(): ByteArray {
        aapsLogger.debug(LTag.PUMPCOMM,
            "UserOptions:" + (System.currentTimeMillis() - danaPump.lastConnection) / 1000 + " s ago"
                + "\ntimeDisplayType24:" + danaPump.timeDisplayType24
                + "\nbuttonScroll:" + danaPump.buttonScrollOnOff
                + "\nbeepAndAlarm:" + danaPump.beepAndAlarm
                + "\nlcdOnTimeSec:" + danaPump.lcdOnTimeSec
                + "\nbacklight:" + danaPump.backlightOnTimeSec
                + "\ndanaRPumpUnits:" + danaPump.units
                + "\nlowReservoir:" + danaPump.lowReservoirRate)
        val request = ByteArray(13)
        request[0] = if (danaPump.timeDisplayType24) 0.toByte() else 1.toByte()
        request[1] = if (danaPump.buttonScrollOnOff) 1.toByte() else 0.toByte()
        request[2] = (danaPump.beepAndAlarm and 0xff).toByte()
        request[3] = (danaPump.lcdOnTimeSec and 0xff).toByte()
        request[4] = (danaPump.backlightOnTimeSec and 0xff).toByte()
        request[5] = (danaPump.selectedLanguage and 0xff).toByte()
        request[6] = (danaPump.units and 0xff).toByte()
        request[7] = (danaPump.shutdownHour and 0xff).toByte()
        request[8] = (danaPump.lowReservoirRate and 0xff).toByte()
        request[9] = (danaPump.cannulaVolume and 0xff).toByte()
        request[10] = (danaPump.cannulaVolume ushr 8 and 0xff).toByte()
        request[11] = (danaPump.refillAmount and 0xff).toByte()
        request[12] = (danaPump.refillAmount ushr 8 and 0xff).toByte()
        return request
    }

    override fun handleMessage(data: ByteArray) {
        val result = intFromBuff(data, 0, 1)
        if (result == 0) {
            aapsLogger.debug(LTag.PUMPCOMM, "Result OK")
            failed = false
        } else {
            aapsLogger.error("Result Error: $result")
            failed = true
        }
    }

    override fun getFriendlyName(): String {
        return "OPTION__SET_USER_OPTION"
    }
}